### PR TITLE
Remove physical expr of ListIndex and ListRange, convert to `array_element` and `array_slice` functions

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #.

Preparation of #9285

### CODE REVIEW

File: datafusion/optimizer/src/analyzer/rewrite_expr.rs [hunk 1]
```diff
@@ -119,6 +121,39 @@ impl TreeNodeRewriter for OperatorToFunctionRewriter {
                 })));
             }
         }
+
+        if let Expr::GetIndexedField(GetIndexedField {
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/rewrite_expr.rs:125]: This is quite clever 👏

File: datafusion/physical-expr/src/expressions/get_indexed_field.rs [hunk 2]
```diff
@@ -42,61 +40,25 @@ use std::{any::Any, sync::Arc};
 pub enum GetFieldAccessExpr {
     /// Named field, For example `struct["name"]`
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/get_indexed_field.rs:41]: 🤔  maybe we should also convert this into a function call as well (as a follow on PR) to remove the entire `GetFieldAccessExpr` physical expr

**Review by alamb — APPROVED** [review]
This is a really nice PR @jayzhan211  -- thank you so much. 

Rewriting the physical expr to use a function is very clever and a very nice reduction in duplication

Screenshot 2024-03-10 at 6 35 12 AM

**Comment by jayzhan211** [comment]
I will file an issue to track the `NamedStructField`. Thanks @alamb

**Pull request merged**